projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,171 @@
+  if (enable) {
+    *channel_enable |= 1 << channel;
+  } else {
+    *channel_enable &= ~(1 << channel);
+  }
+}
+
+// Writes to register given upper and lower bytes.
+static StatusCode prv_setup_register(Ads1015Storage *storage, uint8_t reg, uint8_t msb,
+                                     uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  return i2c_write(storage->i2c_port, storage->i2c_addr + ADS1015_I2C_BASE_ADDRESS,
+                   ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+}
+
+// Reads the register and stores the value in the given array.
+static StatusCode prv_read_register(I2CPort i2c_port, Ads1015Address i2c_addr, uint8_t reg,
+                                    uint8_t *rx_data, size_t rx_len) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, &reg, sizeof(reg)));
+
+  status_ok_or_return(i2c_read(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, rx_data, rx_len));
+  return STATUS_CODE_OK;
+}
+
+// Switches to the given channel by writing to config register.
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         ADS1015_CONFIG_REGISTER_MSB(channel),
+                                         ADS1015_CONFIG_REGISTER_LSB));
+  storage->current_channel = channel;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// Reads and stores the conversion value in storage, and switches to the next enabled channel.
+// Also if there is a callback on a channel, it will be run here.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  Ads1015Storage *storage = context;
+  Ads1015Channel current_channel = storage->current_channel;
+  uint8_t channel_enable = storage->channel_enable;
+  uint8_t read_conv_register[2];
+  prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONV,
+                    read_conv_register, SIZEOF_ARRAY(read_conv_register));
+  // Following line puts the two read bytes into an int16.
+  // 4 LSB's are not part of the result hence the bitshift.
+  storage->channel_readings[current_channel] =
+      ((read_conv_register[0] << 8) | read_conv_register[1]) >> ADS1015_NUM_RESERVED_BITS_CONV_REG;
+  // Runs the users callback if not NULL.
+  if (storage->channel_callback[current_channel] != NULL) {
+    storage->channel_callback[current_channel](current_channel,
+                                               storage->callback_context[current_channel]);
+  }
+  // if more than one channel is enabled go to next channel.
+  if (channel_enable != (1 << current_channel)) {
+    uint8_t temp = channel_enable;
+    temp &= ~((1 << (current_channel + 1)) - 1);
```

Comment:
I think you could just store your channel bitset in your storage instead of all this extra logic. Then you can just remove the first set bit every interrupt.